### PR TITLE
Clean up outliningElementsCollector

### PR DESCRIPTION
**Pull request by andy-ms**
Move stuff out of the big stateful closure into their own functions. There are all pure or at most take an `out` array.

### CODE REVIEW

File: src/services/outliningElementsCollector.ts [hunk 1]
```diff
@@ -1,227 +1,167 @@
...
+        addRegionOutliningSpans(sourceFile, res);
+        return res.sort((span1, span2) => span1.textSpan.start - span2.textSpan.start);
+    }
 
-        walk(sourceFile);
-        gatherRegions();
-        return elements.sort((span1, span2) => span1.textSpan.start - span2.textSpan.start);
+    function addNodeOutliningSpans(sourceFile: SourceFile, cancellationToken: CancellationToken, out: Push<OutliningSpan>): void {
+        let depth = 0;
+        const maxDepth = 39;
```
Comment (uniqueiniquity) [inline src/services/outliningElementsCollector.ts]: Any particular reason we changed from 40 to 39?

File: src/services/outliningElementsCollector.ts [hunk 2]
```diff
@@ -1,227 +1,167 @@
...
-                    const closeBrace = findChildOfKind(n, SyntaxKind.CloseBraceToken, sourceFile);
-                    addOutliningSpan(n, openBrace, closeBrace, autoCollapse(n), /*useFullStart*/ !isArrayLiteralExpression(n.parent));
-                    break;
-                case SyntaxKind.ArrayLiteralExpression:
-                    const openBracket = findChildOfKind(n, SyntaxKind.OpenBracketToken, sourceFile);
-                    const closeBracket = findChildOfKind(n, SyntaxKind.CloseBracketToken, sourceFile);
-                    addOutliningSpan(n, openBracket, closeBracket, autoCollapse(n), /*useFullStart*/ !isArrayLiteralExpression(n.parent));
-                    break;
+            case SyntaxKind.ModuleBlock:
+                return spanForNode(n.parent);
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/outliningElementsCollector.ts]: `spanForNode` closes over `n` and uses that to find tokens. The argument is `hintSpanNode` which is only used for its own span, not for its tokens.

File: src/services/outliningElementsCollector.ts (hunk 1 above)
Comment (andy-ms) [reply src/services/outliningElementsCollector.ts]: Before we counted the SourceFile as depth 1 while now it's depth 0.
I've bumped this back up to 40 so it's a nice round number and changed tests to match.

File: src/services/outliningElementsCollector.ts [hunk 3]
```diff
@@ -1,227 +1,166 @@
...
+    }
 
-        walk(sourceFile);
-        gatherRegions();
-        return elements.sort((span1, span2) => span1.textSpan.start - span2.textSpan.start);
+    function addNodeOutliningSpans(sourceFile: SourceFile, cancellationToken: CancellationToken, out: Push<OutliningSpan>): void {
+        let depthRemaining = 40;
+        sourceFile.forEachChild(function walk(n) {
+            if (depthRemaining === 0) return;
+            cancellationToken.throwIfCancellationRequested();
```
Comment (andy-ms) [inline src/services/outliningElementsCollector.ts]: We are calling this on every single node -- we should come up with a way to call this less often.

**Pull request merged**